### ISSUE TITLE
Weighted Sum of Minecraft Fitness Functions

### ISSUE THREAD

**Issue by schrum2** [schrum2/MM-NEAT#912]
Make a new abstract class called MinecraftWeightedSumFitnessFunction that extends the TimedEvaluationMinecraftFitnessFunction. This abstract class will have a constructor that takes two lists:
- a list of MinecraftFitnessFunctions, and
- a list of doubles (weights)

Given this information, this fitness function will compute fitness scores across all of the given list of MinecraftFitnessFunctions, but take care to separate them out into those that are TimedEvaluationMinecraftFitnessFunctions and those that are not. However, in order to compute the actual fitness score that will be returned, it will take the array of fitness scores returned, multiply each one by its corresponding weight, and add all the results together into a single fitness score.

Basically, this is a way of solving a multiobjective problem by reducing it to a single objective problem. One would expect this to not do very well, but it is a fair baseline to compare NSGA2 and MOME with.

The reason we are making a separate abstract fitness function for this is that the weights will likely need a bit of tuning. The classes that extend this class will just have a constructor that calls the super constructor with a specific list of fitness functions and weights.

**Comment by schrum2** [comment]
I've made the abstract weighted sum fitness, but I still need to extend it with a specific fitness pairing to test it out.

**Comment by schrum2** [comment]
Some classes need to be made that extend the new MinecraftWeightedSumFitnessFunction class. For starters, try one that uses both ChangeBlocksFitness and ChangeCenterOfMass, and make both of the weight values be 0.5. Then try running it with plain MAP Elites. If that works, then make a batch file for it.

**Comment by schrum2** [comment]
Exception resolved

**Comment by JoannaBlatt** [comment]
Was able to run the main method for the ChangeBlockAndChangeCenterOfMassWeightedSumFitness

did not seem to return interesting results. No flying machines.

**Comment by schrum2** [comment]
The next step here is to find a way to log the component fitness scores. I think it will be a bit messy, requiring several methods that return `double[]` to instead return `Pair<double[],double[]>`. I'll give more details on this later.

**Comment by schrum2** [comment]
My most recent commit above takes some steps toward making it possible to track the component scores, but it is not done yet.

**Comment by JoannaBlatt** [comment]
Pure fitness can log weighted sum but map elites crashes when it tries to run

**Comment by schrum2** [comment]
MAP Elites should no longer crash when run with a weighted sum fitness function, though I have not verified this yet. Please check.

However, even though it should not crash, it definitely won't log the desired information yet. You will need to change some of the logging code in MAP Elites itself. In particular, there is a line in MAPElites where this happens:
```
			Float[] elite = ArrayUtils.toObject(archive.getEliteScores());
			final int pseudoGeneration = iterations/individualsPerGeneration;
			archiveLog.log(pseudoGeneration + "\t" + StringUtils.join(elite, "\t").replaceAll("-Infinity", "X"));
```
This code is responsible for logging the objective scores across all bins for the one and only one objective. We need to have an operation similar to this for each component objective of a weighted sum. More generally, the scores we are interested in logging will be within the `otherStats` field of each `Score` instance in a bin, and all scores should have the same number of entries in this array. So, there needs to be a separate log and plot file for each index in `otherStats`. Similarly, there needs to be a method in `Archive` like `getEliteScores()`, except instead of getting the fitness score for each bin, it gets the score in a specific `otherStats` index for each bin, as in:
```
getOtherStatsScores(int index)
```
Given what this returns, the data can be logged and plotted the same way.

MAPElites also uses this code to compute the max fitness:
```
			Float maximumFitness = StatisticsUtilities.maximum(elite);
```
and this code to compute QD scores with respect to the fitness scores:
```
			final double qdScore = calculateQDScore(elite);
```

If you were to simply send an array of otherStats values to these methods, then you would compute both a max score in the other stat, as well as a QD score across the otherStat. This would all need to be logged and plotted with distinct files, but it mostly comes down to copying what is already there with just small tweaks. A method can probably be extracted that creates plt files that differ only in the name of the log file they are plotting, and then the logging commands themselves can just be different log files (though you will need an array of log files for each of the otherStats values).

**Comment by JoannaBlatt** [comment]
I might have created an issue with pseudoArchive setting due to passing an arraylist of logs that are not set up there. This may have backwards compatibility issues.

**Comment by JoannaBlatt** [comment]
Currently logs. Handling in setupLogging currently similar to momeLogs, but could probably be handled differently. 
Issues with negative scores are present. It should probably be accounted for elsewhere.

**Comment by schrum2** [comment]
Running tests now, but all scores should be in the range 0.0 to 1.0 now. Won't be able to check results until Monday though.

**Comment by schrum2** [comment]
Scores to range from 0 to 1 in a separate test that I did, so I'm closing this issue. However, there is a correction for negative scores needed with QD calculations, and that is tracked in this new issue #943
